File: sdk/typescript/packages/hmac/src/env-loader.ts
import type {
  HmacClientIdentity,
  HmacConfig,
  HmacTargetConfig,
  SignedHeadersConfig,
} from "./config.js";
import {
  DEFAULT_TIMESTAMP_TOLERANCE_SECONDS,
  defaultSignedHeadersConfig,
} from "./config.js";

/**
 * Try to load dotenv if it's available as a peer dependency.
 * This is a best-effort load — if dotenv is not installed, we skip silently.
 * Uses dynamic import() for ESM compatibility.
 */
async function tryLoadDotenv(): Promise<void> {
  try {
    const dotenv = await import("dotenv");
    dotenv.config();
  } catch {
    // dotenv is an optional peer dependency — skip if not available
  }
}

function parseBool(value: string | undefined, defaultValue: boolean): boolean {
  if (value === undefined) return defaultValue;
  return ["true", "1", "yes"].includes(value.toLowerCase());
}

/**
 * Load HmacConfig from environment variables.
 *
 * Supports the following variables (prefix default: "HARDEN_HMAC_"):
 *   {prefix}SHARED_SECRET_BASE64 — global shared secret
 *   {prefix}TIMESTAMP_TOLERANCE_SECONDS — global tolerance
 *   {prefix}SIGNED_HEADERS__INCLUDE_AUTHORIZATION — true/false
 *   {prefix}SIGNED_HEADERS__INCLUDE_X_HEADERS — true/false
 *   {prefix}TARGETS__{NAME}__BASE_URL — per-target base URL
 *   {prefix}TARGETS__{NAME}__SHARED_SECRET — per-target secret
 *   {prefix}TARGETS__{NAME}__TIMESTAMP_TOLERANCE_SECONDS — per-target tolerance
 *
 * If dotenv is installed as a peer dependency, .env file variables are loaded first.
 *
 * @param prefix - Environment variable prefix. Default: "HARDEN_HMAC_".
 * @param env - Optional env object for testing. Defaults to process.env.
 * @returns Parsed HmacConfig.
 */
/**
 * Parse an integer from a string, returning the default if the value is
 * undefined, empty, or not a valid integer.
 */
function safeParseInt(value: string | undefined, defaultValue: number): number {
  if (value === undefined || value === "") return defaultValue;
  const parsed = parseInt(value, 10);
  if (isNaN(parsed)) return defaultValue;
  return parsed;
}

export async function fromEnv(
  prefix: string = "HARDEN_HMAC_",
  env?: Record<string, string | undefined>
): Promise<HmacConfig> {
  if (!env) {
    await tryLoadDotenv();
    env = process.env;
  }

  const upperPrefix = prefix.toUpperCase();

  // Extract all prefixed vars
  const prefixed: Record<string, string> = {};
  for (const [key, value] of Object.entries(env)) {
    if (key && value !== undefined && key.toUpperCase().startsWith(upperPrefix)) {
      const suffix = key.substring(upperPrefix.length).toUpperCase();
      prefixed[suffix] = value;
    }
  }

  // Parse global settings
  const sharedSecretBase64 = prefixed["SHARED_SECRET_BASE64"] ?? "";
  const timestampToleranceSeconds = safeParseInt(
    prefixed["TIMESTAMP_TOLERANCE_SECONDS"],
    DEFAULT_TIMESTAMP_TOLERANCE_SECONDS
  );

  // Parse signed headers
  const signedHeaders: SignedHeadersConfig = {
    ...defaultSignedHeadersConfig(),
    includeAuthorization: parseBool(
      prefixed["SIGNED_HEADERS__INCLUDE_AUTHORIZATION"],
      true
    ),
    includeXHeaders: parseBool(
      prefixed["SIGNED_HEADERS__INCLUDE_X_HEADERS"],
      true
    ),
  };

  // Parse targets: keys matching TARGETS__{NAME}__{FIELD}
  const targetFields: Record<string, Record<string, string>> = {};
  const targetPrefix = "TARGETS__";
  for (const [key, value] of Object.entries(prefixed)) {
    if (!key.startsWith(targetPrefix)) continue;
    const rest = key.substring(targetPrefix.length);
    const separatorIndex = rest.indexOf("__");
    if (separatorIndex === -1) continue;
    const targetName = rest.substring(0, separatorIndex).toLowerCase().replace(/_/g, "-");
    const fieldName = rest.substring(separatorIndex + 2).toUpperCase();
    if (!targetFields[targetName]) {
      targetFields[targetName] = {};
    }
    targetFields[targetName][fieldName] = value;
  }

  const targets: Record<string, HmacTargetConfig> = {};
  for (const [name, fields] of Object.entries(targetFields)) {
    let targetSignedHeaders: SignedHeadersConfig | undefined;
    if (
      fields["SIGNED_HEADERS__INCLUDE_AUTHORIZATION"] !== undefined ||
      fields["SIGNED_HEADERS__INCLUDE_X_HEADERS"] !== undefined
    ) {
      targetSignedHeaders = {
        ...defaultSignedHeadersConfig(),
        includeAuthorization: parseBool(
          fields["SIGNED_HEADERS__INCLUDE_AUTHORIZATION"],
          true
        ),
        includeXHeaders: parseBool(
          fields["SIGNED_HEADERS__INCLUDE_X_HEADERS"],
          true
        ),
      };
    }

    let targetTimestampTolerance: number | undefined;
    if (fields["TIMESTAMP_TOLERANCE_SECONDS"] !== undefined) {
      const parsed = parseInt(fields["TIMESTAMP_TOLERANCE_SECONDS"], 10);
      targetTimestampTolerance = isNaN(parsed) ? undefined : parsed;
    }

    targets[name] = {
      baseUrl: fields["BASE_URL"] ?? "",
      sharedSecret: fields["SHARED_SECRET"] ?? "",
      signedHeaders: targetSignedHeaders,
      timestampToleranceSeconds: targetTimestampTolerance,
    };
  }

  // Parse clients: keys matching CLIENTS__{NAME}__{FIELD}
  const clientFields: Record<string, Record<string, string>> = {};
  const clientPrefix = "CLIENTS__";
  for (const [key, value] of Object.entries(prefixed)) {
    if (!key.startsWith(clientPrefix)) continue;
    const rest = key.substring(clientPrefix.length);
    const separatorIndex = rest.indexOf("__");
    if (separatorIndex === -1) continue;
    const clientName = rest.substring(0, separatorIndex).toLowerCase().replace(/_/g, "-");
    const fieldName = rest.substring(separatorIndex + 2).toUpperCase();
    if (!clientFields[clientName]) {
      clientFields[clientName] = {};
    }
    clientFields[clientName][fieldName] = value;
  }

  const clients: Record<string, HmacClientIdentity> = {};
  for (const [name, fields] of Object.entries(clientFields)) {
    clients[name] = {
      sharedSecret: fields["SHARED_SECRET"] ?? "",
    };
  }

  return {
    sharedSecretBase64,
    targets: Object.keys(targets).length > 0 ? targets : undefined,
    clients: Object.keys(clients).length > 0 ? clients : undefined,
    signedHeaders,
    timestampToleranceSeconds,
  };
}

File: sdk/python/src/hardenlabs_hmac/env_loader.py
"""Environment variable parsing for HardenHMAC configuration."""

from __future__ import annotations

import os
from typing import Any


def _try_load_dotenv() -> None:
    """Attempt to load .env file if python-dotenv is available."""
    try:
        from dotenv import load_dotenv

        load_dotenv()
    except ImportError:
        pass


def _parse_bool(value: str) -> bool:
    """Parse a boolean string value."""
    return value.lower() in ("true", "1", "yes")


def load_config_from_env(
    prefix: str = "HARDEN_HMAC_",
    env: dict[str, str] | None = None,
) -> "HmacConfig":
    """Parse environment variables into an HmacConfig.

    If python-dotenv is installed and no explicit env dict is provided,
    .env file variables are loaded into os.environ first.

    Args:
        prefix: Environment variable prefix.
        env: Optional dict of environment variables (for testing).
             If None, uses os.environ after loading .env.

    Returns:
        Parsed HmacConfig.
    """
    from hardenlabs_hmac.config import (
        HmacClientIdentity,
        HmacConfig,
        HmacTargetConfig,
        SignedHeadersConfig,
    )

    if env is None:
        _try_load_dotenv()
        env = dict(os.environ)

    upper_prefix = prefix.upper()

    # Extract all env vars with the prefix
    prefixed: dict[str, str] = {}
    for key, value in env.items():
        if key.upper().startswith(upper_prefix):
            # Strip prefix and normalize to uppercase
            suffix = key[len(upper_prefix):]
            prefixed[suffix.upper()] = value

    # Parse global settings
    shared_secret = prefixed.get("SHARED_SECRET_BASE64", "")
    try:
        timestamp_tolerance = int(prefixed.get("TIMESTAMP_TOLERANCE_SECONDS", "30"))
    except ValueError:
        timestamp_tolerance = 30

    # Parse signed headers
    include_auth = _parse_bool(
        prefixed.get("SIGNED_HEADERS__INCLUDE_AUTHORIZATION", "true")
    )
    include_x = _parse_bool(
        prefixed.get("SIGNED_HEADERS__INCLUDE_X_HEADERS", "true")
    )
    signed_headers = SignedHeadersConfig(
        include_authorization=include_auth,
        include_x_headers=include_x,
    )

    # Parse targets: keys matching TARGETS__{NAME}__{FIELD}
    targets: dict[str, dict[str, Any]] = {}
    target_prefix = "TARGETS__"
    for key, value in prefixed.items():
        if not key.startswith(target_prefix):
            continue
        rest = key[len(target_prefix):]
        parts = rest.split("__", 1)
        if len(parts) != 2:
            continue
        target_name = parts[0].lower().replace("_", "-")
        field_name = parts[1].upper()
        if target_name not in targets:
            targets[target_name] = {}
        targets[target_name][field_name] = value

    target_configs: dict[str, HmacTargetConfig] = {}
    for name, fields in targets.items():
        target_signed_headers: SignedHeadersConfig | None = None
        if "SIGNED_HEADERS__INCLUDE_AUTHORIZATION" in fields or "SIGNED_HEADERS__INCLUDE_X_HEADERS" in fields:
            target_signed_headers = SignedHeadersConfig(
                include_authorization=_parse_bool(
                    fields.get("SIGNED_HEADERS__INCLUDE_AUTHORIZATION", "true")
                ),
                include_x_headers=_parse_bool(
                    fields.get("SIGNED_HEADERS__INCLUDE_X_HEADERS", "true")
                ),
            )

        target_tolerance: int | None = None
        if "TIMESTAMP_TOLERANCE_SECONDS" in fields:
            try:
                target_tolerance = int(fields["TIMESTAMP_TOLERANCE_SECONDS"])
            except ValueError:
                target_tolerance = None

        target_configs[name] = HmacTargetConfig(
            base_url=fields.get("BASE_URL", ""),
            shared_secret=fields.get("SHARED_SECRET", ""),
            signed_headers=target_signed_headers,
            timestamp_tolerance_seconds=target_tolerance,
        )

    # Parse clients: keys matching CLIENTS__{NAME}__{FIELD}
    clients: dict[str, dict[str, str]] = {}
    client_prefix = "CLIENTS__"
    for key, value in prefixed.items():
        if not key.startswith(client_prefix):
            continue
        rest = key[len(client_prefix):]
        parts = rest.split("__", 1)
        if len(parts) != 2:
            continue
        client_name = parts[0].lower().replace("_", "-")
        field_name = parts[1].upper()
        if client_name not in clients:
            clients[client_name] = {}
        clients[client_name][field_name] = value

    client_configs: dict[str, HmacClientIdentity] = {}
    for name, fields in clients.items():
        client_configs[name] = HmacClientIdentity(
            shared_secret=fields.get("SHARED_SECRET", ""),
        )

    return HmacConfig(
        shared_secret_base64=shared_secret,
        targets=target_configs,
        clients=client_configs,
        signed_headers=signed_headers,
        timestamp_tolerance_seconds=timestamp_tolerance,
    )
